**复习：**在前面我们已经学习了Pandas基础，知道利用Pandas读取csv数据的增删查改，今天我们要学习的就是**探索性数据分析**，主要介绍如何利用Pandas进行排序、算术计算以及计算描述函数describe()的使用。

# 1 第一章：探索性数据分析

#### 开始之前，导入numpy、pandas包和数据

In [1]:
#加载所需的库
import numpy as np
import pandas as pd

In [2]:
#载入之前保存的train_chinese.csv数据，关于泰坦尼克号的任务，我们就使用这个数据
df=pd.read_csv('train_chinese.csv')
df.head(3)

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价   客舱 登船港口  
0    male  22.0       1       0         A/5 21171   7.2500  NaN    S  
1  female  38.0       1       0          PC 17599  71.2833  C85    C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250  NaN    S

### 1.6 了解你的数据吗？
教材《Python for Data Analysis》第五章

#### 1.6.1 任务一：利用Pandas对示例数据进行排序，要求升序

In [7]:
# 具体请看《利用Python进行数据分析》第五章 排序和排名 部分
#自己构建一个都为数字的DataFrame数据
d=pd.DataFrame(data=np.arange(8).reshape((2,4)),index=[2,1],columns=['d','a','b','c'])
d

d  a  b  c
2  0  1  2  3
1  4  5  6  7

【代码解析】

pd.DataFrame() ：创建一个DataFrame对象 

np.arange(8).reshape((2, 4)) : 生成一个二维数组（2*4）,第一列：0，1，2，3 第二列：4，5，6，7

index=['2, 1] ：DataFrame 对象的索引列

columns=['d', 'a', 'b', 'c'] ：DataFrame 对象的索引行

【问题】：大多数时候我们都是想根据列的值来排序,所以将你构建的DataFrame中的数据根据某一列，升序排列

In [11]:
#回答代码
d.sort_values(by='c', ascending=False)#根据c列降序

d  a  b  c
1  4  5  6  7
2  0  1  2  3

【思考】通过书本你能说出Pandas对DataFrame数据的其他排序方式吗？

【总结】下面将不同的排序方式做一个总结

1.让行索引升序排序

In [12]:
#代码
d.sort_index()#默认升序ascending=true，默认axis=0

d  a  b  c
1  4  5  6  7
2  0  1  2  3

2.让列索引升序排序


In [13]:
#代码
d.sort_index(axis=1,ascending=True)

a  b  c  d
2  1  2  3  0
1  5  6  7  4

3.让列索引降序排序

In [14]:
#代码
d.sort_index(axis=1,ascending=False)

d  c  b  a
2  0  3  2  1
1  4  7  6  5

4.让任选两列数据同时降序排序

In [17]:
#代码
d.sort_values(by=['a', 'b'],ascending=False)

d  a  b  c
1  4  5  6  7
2  0  1  2  3

#### 1.6.2 任务二：对泰坦尼克号数据（trian.csv）按票价和年龄两列进行综合排序（降序排列），从这个数据中你可以分析出什么？


在开始我们已经导入了train_chinese.csv数据，而且前面我们也学习了导入数据过程，根据上面学习，我们直接对目标列进行排序即可
head(20) : 读取前20条数据



In [23]:
#代码
df=pd.read_csv('train.csv')
df.sort_values(by=['Fare', 'Age'], ascending=False).head(3)

PassengerId  Survived  Pclass                                Name  \
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C

【思考】排序后，如果我们仅仅关注年龄和票价两列。根据常识我知道发现票价越高的应该客舱越好，所以我们会明显看出，票价前20的乘客中存活的有14人，这是相当高的一个比例，那么我们后面是不是可以进一步分析一下票价和存活之间的关系，年龄和存活之间的关系呢？当你开始发现数据之间的关系了，数据分析就开始了。

当然，这只是我的想法，你还可以有更多想法，欢迎写在你的学习笔记中。

**多做几个数据的排序**

In [32]:
#代码
df.sort_values(by=['Pclass'], ascending=True).head(3)

PassengerId  Survived  Pclass                            Name     Sex  \
445          446         1       1       Dodge, Master. Washington    male   
310          311         1       1  Hays, Miss. Margaret Bechstein  female   
309          310         1       1  Francatelli, Miss. Laura Mabel  female   

      Age  SibSp  Parch    Ticket     Fare Cabin Embarked  
445   4.0      0      2     33638  81.8583   A34        S  
310  24.0      0      0     11767  83.1583   C54        C  
309  30.0      0      0  PC 17485  56.9292   E36        C

In [4]:
#写下你的思考
仓位等级升序排列，前20个都是存活的,后20个存活5人，说明仓位越好的人存活率越高。这个思路和票价是一样的。
另外Age有很多是缺测的，而没有存活的缺测比例更高。用年纪作为存活率的一个指标是否具有幸存者偏差。


#### 1.6.3 任务三：利用Pandas进行算术计算，计算两个DataFrame数据相加结果

In [10]:
# 具体请看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分

#自己构建两个都为数字的DataFrame数据

"""
我们举了一个例子：
frame1_a = pd.DataFrame(np.arange(9.).reshape(3, 3),
                     columns=['a', 'b', 'c'],
                     index=['one', 'two', 'three'])
frame1_b = pd.DataFrame(np.arange(12.).reshape(4, 3),
                     columns=['a', 'e', 'c'],
                     index=['first', 'one', 'two', 'second'])
frame1_a
"""

In [38]:
#代码
d1=pd.DataFrame(data=np.arange(9.).reshape(3, 3),columns=['a', 'b', 'c'],index=['one', 'two', 'three'])
d2=pd.DataFrame(data=np.arange(12.).reshape(4, 3),columns=['a', 'e', 'c'],index=['first', 'one', 'two', 'second'])
print(d1)
print(d2)

         a    b    c
one    0.0  1.0  2.0
two    3.0  4.0  5.0
three  6.0  7.0  8.0
          a     e     c
first   0.0   1.0   2.0
one     3.0   4.0   5.0
two     6.0   7.0   8.0
second  9.0  10.0  11.0


将frame_a和frame_b进行相加


In [41]:
#代码
d1+d2#对应的行与列相加

a   b     c   e
first   NaN NaN   NaN NaN
one     3.0 NaN   7.0 NaN
second  NaN NaN   NaN NaN
three   NaN NaN   NaN NaN
two     9.0 NaN  13.0 NaN

【提醒】两个DataFrame相加后，会返回一个新的DataFrame，对应的行和列的值会相加，没有对应的会变成空值NaN。<br>
当然，DataFrame还有很多算术运算，如减法，除法等，有兴趣的同学可以看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分，多在网络上查找相关学习资料。

第三节的任务三也可以和原始数据结合。例如：df_chinese['家族人数']=df_chinese['兄弟姐妹个数']+df_chinese['父母子女个数']。这样的两个dataframe可以得到新数据

#### 1.6.4 任务四：通过泰坦尼克号数据如何计算出在船上最大的家族有多少人？

In [64]:
'''
还是用之前导入的chinese_train.csv如果我们想看看在船上，最大的家族有多少人（‘兄弟姐妹个数’+‘父母子女个数’），我们该怎么做呢？
'''


'\n还是用之前导入的chinese_train.csv如果我们想看看在船上，最大的家族有多少人（‘兄弟姐妹个数’+‘父母子女个数’），我们该怎么做呢？\n'

In [56]:
#代码
df_chinese=pd.read_csv('train.csv')
df_chinese=pd.read_csv('train_chinese.csv')
df_chinese['家族人数']=df_chinese['兄弟姐妹个数']+df_chinese['父母子女个数']
df_chinese.sort_values(by=['家族人数'],ascending=False).head(3)#最大家族有10人

乘客ID  是否幸存  仓位等级                            姓名      性别  年龄  兄弟姐妹个数  \
180   181     0     3  Sage, Miss. Constance Gladys  female NaN       8   
846   847     0     3      Sage, Mr. Douglas Bullen    male NaN       8   
792   793     0     3       Sage, Miss. Stella Anna  female NaN       8   

     父母子女个数      船票信息     票价   客舱 登船港口  家族人数  
180       2  CA. 2343  69.55  NaN    S    10  
846       2  CA. 2343  69.55  NaN    S    10  
792       2  CA. 2343  69.55  NaN    S    10

In [59]:
max(df_chinese['兄弟姐妹个数'] + df_chinese['父母子女个数'])

10

【提醒】我们只需找出”兄弟姐妹个数“和”父母子女个数“之和最大的数，当然你还可以想出很多方法和思考角度，欢迎你来说出你的看法。

**多做几个数据的相加，看看你能分析出什么？**

In [60]:
#代码
max((df_chinese['兄弟姐妹个数'] + df_chinese['父母子女个数'])*df_chinese['是否幸存'])

6

这样计算得到的是幸存的人所在家族的最多人口数

In [63]:
sum(df_chinese['是否幸存'])#幸存的总人数

342

#### 1.6.5 任务五：学会使用Pandas describe()函数查看数据基本统计信息

In [13]:
#(1) 关键知识点示例做一遍（简单数据）
# 具体请看《利用Python进行数据分析》第五章 汇总和计算描述统计 部分

#自己构建一个有数字有空值的DataFrame数据


"""
我们举了一个例子：
frame2 = pd.DataFrame([[1.4, np.nan], 
                       [7.1, -4.5],
                       [np.nan, np.nan], 
                       [0.75, -1.3]
                      ], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])
frame2

"""

In [65]:
#代码
frame2 = pd.DataFrame([[1.4, np.nan], [7.1, -4.5],[np.nan, np.nan], [0.75, -1.3]], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])
frame2

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

调用 describe 函数，观察frame2的数据基本信息

In [66]:
#代码
frame2.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

#### 1.6.6 任务六：分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据，你能发现什么？

In [ ]:
'''
看看泰坦尼克号数据集中 票价 这列数据的基本统计数据
'''

In [67]:
#代码
df_chinese['票价'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: 票价, dtype: float64

【思考】从上面数据我们可以看出，试试在下面写出你的看法。然后看看我们给出的答案。

当然，答案只是我的想法，你还可以有更多想法，欢迎写在你的学习笔记中。

统计了票价的一共有891人，最贵的花了512，最便宜的花了0，平均值为32，方差为49，属于挺大的。25%的人票价低于7.9,50%的人票价低于14.45，75%的人票价低于31。

**多做几个组数据的统计，看看你能分析出什么？**

In [68]:
# 写下你的其他分析
df_chinese['仓位等级'].describe()

count    891.000000
mean       2.308642
std        0.836071
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max        3.000000
Name: 仓位等级, dtype: float64

【思考】有更多想法，欢迎写在你的学习笔记中。

【总结】本节中我们通过Pandas的一些内置函数对数据进行了初步统计查看，这个过程最重要的不是大家得掌握这些函数，而是看懂从这些函数出来的数据，构建自己的数据分析思维，这也是第一章最重要的点，希望大家学完第一章能对数据有个基本认识，了解自己在做什么，为什么这么做，后面的章节我们将开始对数据进行清洗，进一步分析。